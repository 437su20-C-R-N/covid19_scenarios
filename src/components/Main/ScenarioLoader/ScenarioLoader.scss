@use "sass:map";

@import '../../../styles/variables';
@import '../../../styles/bootstrap-variables';

$lg: map.get($grid-breakpoints, 'lg');
$md: map.get($grid-breakpoints, 'md');

#scenario-loader-modal {
  @media screen and (max-width: $md) {
    .modal-content {
      position: fixed;
      top: 0;
      right: 0;
      left: 0;
      overflow-y: scroll;
    }
  }

  @media screen and (min-width: $md) {
    width: 90vw;
    max-width: 1200px;
  }
}

.modal-content {
  background-color: $body-bg;
}

.scenario-loader-card {
  &.card {
    background-color: $white;
    border-width: 0;
    box-shadow: 1px 1px 3px 3px $gray-400;
  }

  @media screen and (max-width: $md) {
    height: 75vh;
    margin-bottom: 4px;
  }

  @media screen and (min-width: $md) {
    height: 85vh;
  }
}

.scenario-loader-list-container {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
}

.scenario-loader-list-input-group,
.scenario-loader-list {
  display: flex;
  flex-grow: 1;
}

.scenario-loader-list-input-group {
  flex-shrink: 0;
}

.scenario-loader-list {
  flex-direction: column;
  overflow: scroll; /* other browsers */
  overflow: overlay; /* Chrome */
}

.scenario-loader-list-item {
  cursor: pointer;

  ::selection,
  :active,
  :hover {
    background-color: lighten($primary, 33%);
  }
}

.scenario-loader-uploader-error-list-container {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
}

.scenario-loader-uploader-error-list {
  height: 100%;
  overflow-y: scroll;
}

.preset-loader-button {
  @media screen and (max-width: $lg) {
    width: 100%;
  }
}

$upload-zone-fg: $gray-500;
$upload-zone-bg: $gray-100;

.upload-zone {
  min-height: 100px;
  cursor: pointer;
  background-color: $upload-zone-bg;
  border: 5px $upload-zone-fg dashed;

  &.upload-zone-active {
    border-style: dotted;
  }
}

.upload-icon {
  fill: $upload-zone-fg;
}
