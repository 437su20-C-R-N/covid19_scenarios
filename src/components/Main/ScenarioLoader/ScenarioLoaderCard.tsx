import React, { ReactNode } from 'react'

import { Card, CardBody, CardHeader } from 'reactstrap'

export interface ScenarioLoaderCardProps {
  header: ReactNode
  children: ReactNode
}

function ScenarioLoaderCard({ header, children }: React.PropsWithChildren<ScenarioLoaderCardProps>) {
  return (
    <Card className="scenario-loader-card">
      <CardHeader className="py-2">
        <h3 className="px-2">{header}</h3>
      </CardHeader>

      <CardBody>{children}</CardBody>
    </Card>
  )
}

export { ScenarioLoaderCard }
