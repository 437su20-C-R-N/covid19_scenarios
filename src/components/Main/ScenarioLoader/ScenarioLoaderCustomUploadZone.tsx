import React from 'react'

import { DropEvent, useDropzone } from 'react-dropzone'

import { useTranslation } from 'react-i18next'

export interface ScenarioLoaderCustomUploadZoneProps {
  onDrop<T extends File>(acceptedFiles: T[], rejectedFiles: T[], event: DropEvent): void
}

export function ScenarioLoaderCustomUploadZone({ onDrop }: ScenarioLoaderCustomUploadZoneProps) {
  const { getRootProps, getInputProps, isDragActive } = useDropzone({ onDrop, multiple: false })
  const { t } = useTranslation()
  return (
    <div>
      <div {...getRootProps()}>
        <input type="file" {...getInputProps()} />
        {isDragActive ? (
          <p>{t('Drop the files here ...')}</p>
        ) : (
          <p>{t("Drag n' drop some files here, or click to select files")}</p>
        )}
      </div>
    </div>
  )
}
