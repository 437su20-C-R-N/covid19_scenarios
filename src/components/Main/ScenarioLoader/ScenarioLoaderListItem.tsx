import React from 'react'

import { NavLink } from 'reactstrap'

import './ScenarioLoader.scss'

export interface ScenarioOption {
  label: string
  value: string
}

export interface ScenarioLoaderListItemProps {
  option: ScenarioOption
  onItemClick: (id: string) => void
}

export function ScenarioLoaderListItem({ option, onItemClick }: ScenarioLoaderListItemProps) {
  return (
    <div className="scenario-loader-row">
      <NavLink
        data-testid="PresetLoaderDialogRowLink"
        onClick={(event) => {
          event.preventDefault()
          onItemClick(option.value)
        }}
      >
        {option.label}
      </NavLink>
    </div>
  )
}
