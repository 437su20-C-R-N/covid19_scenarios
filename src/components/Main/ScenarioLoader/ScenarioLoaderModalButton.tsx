import React, { HTMLProps, useState, useMemo } from 'react'

import { useTranslation } from 'react-i18next'
import type { AnyAction } from 'typescript-fsa'
import { Button } from 'reactstrap'

import type { SeverityDistributionDatum } from '../../../algorithms/types/Param.types'
import type { ScenarioOption } from './ScenarioOption'

import { ScenarioLoader } from './ScenarioLoader'

import './ScenarioLoader.scss'

interface ScenarioLoaderModalButtonProps extends HTMLProps<HTMLButtonElement> {}

export function ScenarioLoaderModalButton({ scenarioNames }: ScenarioLoaderModalButtonProps) {
  const { t } = useTranslation()
  const [showDialog, setShowDialog] = useState(false)

  const scenarioOptions = useMemo(stringsToOptions(scenarioNames), [scenarioNames])

  return (
    <>
      <Button
        className="scenario-loader-button"
        color="primary"
        size="sm"
        data-testid="PresetLoaderButton"
        onClick={() => setShowDialog(true)}
      >
        {t('Load')}
      </Button>

      <ScenarioLoader
        scenarioOptions={scenarioOptions}
        onScenarioSelect={(id: string) => {
          onScenarioSelect(id)
          setShowDialog(false)
        }}
        onClose={() => {
          setShowDialog(false)
        }}
        setSeverity={setSeverity}
        scenarioDispatch={scenarioDispatch}
        visible={showDialog}
      />
    </>
  )
}
