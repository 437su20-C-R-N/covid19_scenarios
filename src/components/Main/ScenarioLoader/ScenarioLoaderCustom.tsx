import React from 'react'

import type { AnyAction } from 'typescript-fsa'

import type { SeverityDistributionDatum } from '../../../algorithms/types/Param.types'

import { readFile } from '../../../helpers/readFile'

import { setStateData } from '../state/actions'
import { deserialize } from '../state/serialize'

import { ScenarioLoaderCustomUploadZone } from './ScenarioLoaderCustomUploadZone'

export interface ScenarioLoaderCustomProps {
  setSeverity(severity: SeverityDistributionDatum[]): void
  scenarioDispatch(action: AnyAction): void
}

export function ScenarioLoaderCustom({ scenarioDispatch, setSeverity }: ScenarioLoaderCustomProps) {
  async function processParamJson(file: File) {
    const str = await readFile(file)
    const params = deserialize(str)

    if (!params) {
      throw new Error('Failed to deserialize parameters from JSON file')
    }

    scenarioDispatch(
      setStateData({
        current: params.scenarioName,
        data: params.scenario,
        ageDistribution: params.ageDistribution,
      }),
    )
    setSeverity(params.severity)
  }

  async function onDrop(files: File[]) {
    if (files.length > 0) {
      await processParamJson(files[0])
    }
  }

  return <ScenarioLoaderCustomUploadZone onDrop={onDrop} />
}
