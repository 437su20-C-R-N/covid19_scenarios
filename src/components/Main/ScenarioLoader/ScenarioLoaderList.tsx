import React, { HTMLProps, KeyboardEvent, useEffect, useState } from 'react'

import { useTranslation } from 'react-i18next'
import { MdClear } from 'react-icons/md'
import { Button, Col, Input, InputGroup, InputGroupAddon, Row } from 'reactstrap'
import { useDebouncedCallback } from 'use-debounce'

import { ScenarioLoaderListItem, ScenarioOption } from './ScenarioLoaderListItem'

import './ScenarioLoader.scss'

const DEBOUNCE_DELAY = 500

export interface ScenarioLoaderListProps extends HTMLProps<HTMLDivElement> {
  items: ScenarioOption[]
  onScenarioSelect(scenario: string): void
}

export function ScenarioLoaderList({ items, onScenarioSelect, hidden }: ScenarioLoaderListProps) {
  const { t } = useTranslation()
  const [searchTerm, setSearchTerm] = useState('')
  const [filteredRows, setFilteredRows] = useState(items)

  const executeFilter = (term: string) => {
    const hasSearchTerm = term.length > 0
    const filtered = hasSearchTerm
      ? items.filter((item) => item.label.toLowerCase().includes(term.toLowerCase()))
      : items

    setFilteredRows(filtered)
  }
  const [executeFilterDebounced] = useDebouncedCallback(executeFilter, DEBOUNCE_DELAY)

  useEffect(() => {
    if (!hidden) {
      /* TODO for some reason the useRef way doesn't work */
      setTimeout(() => {
        const element: HTMLInputElement | null = document.getElementById(
          'preset-loader-dialog-input',
        ) as HTMLInputElement

        if (element) {
          element.focus()
          element.select()
        }
      }, 0)
    }
  }, [hidden])

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target

    setSearchTerm(value)
    executeFilterDebounced(value)
  }

  const onKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    const isEnterKey = event.keyCode === 13

    if (isEnterKey) {
      executeFilter(searchTerm)
    }
  }

  return (
    <>
      <Row className="mb-2">
        <Col>
          <InputGroup>
            <Input
              className="input-search-scenario"
              name="search-scenario"
              data-testid="PresetLoaderDialogInput"
              placeholder={t('Search')}
              onChange={onChange}
              value={searchTerm}
              autoComplete="off"
              onKeyDown={onKeyDown}
            />
            <InputGroupAddon addonType="append">
              <Button
                color="secondary"
                className="btn-search-scenario"
                data-testid="PresetLoaderDialogClearButton"
                disabled={searchTerm === ''}
                onClick={() => {
                  setSearchTerm('')
                  executeFilter('')
                }}
              >
                <MdClear />
              </Button>
            </InputGroupAddon>
          </InputGroup>
        </Col>
      </Row>

      <Row className="h-100">
        <Col className="h-100">
          <div className="preset-loader-dialog-table-container">
            {filteredRows.map((item) => (
              <ScenarioLoaderListItem key={item.value} option={item} onItemClick={onScenarioSelect} />
            ))}
          </div>
        </Col>
      </Row>
    </>
  )
}
