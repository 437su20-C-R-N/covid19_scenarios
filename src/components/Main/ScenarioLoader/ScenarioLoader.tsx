import React from 'react'

import { useTranslation } from 'react-i18next'
import { Modal, ModalBody, ModalHeader, Col, Row } from 'reactstrap'

import type { AnyAction } from 'typescript-fsa'

import type { SeverityDistributionDatum } from '../../../algorithms/types/Param.types'

import { ScenarioOption } from './ScenarioLoaderListItem'

import { ScenarioLoaderCustom } from './ScenarioLoaderCustom'
import { ScenarioLoaderList } from './ScenarioLoaderList'

import './ScenarioLoader.scss'

export interface ScenarioLoaderProps {
  visible: boolean
  scenarioOptions: ScenarioOption[]
  onScenarioSelect: (id: string) => void
  onClose: () => void
  setSeverity(severity: SeverityDistributionDatum[]): void
  scenarioDispatch(action: AnyAction): void
}

export function ScenarioLoader({
  scenarioOptions,
  visible,
  onScenarioSelect,
  onClose,
  setSeverity,
  scenarioDispatch,
}: ScenarioLoaderProps) {
  const { t } = useTranslation()

  return (
    <Modal
      id={'scenario-loader-modal'}
      centered
      isOpen={visible}
      toggle={onClose}
      fade={false}
      data-testid="PresetLoaderDialog"
    >
      <ModalHeader toggle={onClose}>{t(`Change scenario`)}</ModalHeader>

      <ModalBody>
        <Row className="scenario-loader-modal">
          <Col className="h-100" md={6}>
            <ScenarioLoaderList items={scenarioOptions} onScenarioSelect={onScenarioSelect} hidden={!visible} />
          </Col>

          <Col md={6}>
            <ScenarioLoaderCustom setSeverity={setSeverity} scenarioDispatch={scenarioDispatch} />
          </Col>
        </Row>
      </ModalBody>
    </Modal>
  )
}
