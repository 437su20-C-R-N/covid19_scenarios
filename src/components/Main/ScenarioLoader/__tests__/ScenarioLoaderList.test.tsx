import React from 'react'

import { noop } from 'lodash'
import { cleanup, fireEvent, render, RenderResult, wait } from '@testing-library/react'
import { AnyAction } from 'typescript-fsa'

import type { SeverityDistributionDatum } from '../../../../algorithms/types/Param.types'

import type { ScenarioOption } from '../ScenarioLoaderListItem'

import { ScenarioLoader } from '../ScenarioLoader'

const DATA: ScenarioOption[] = [
  {
    label: 'First row',
    value: '1',
  },
  {
    label: 'Second row',
    value: '2',
  },
  {
    label: 'Third row',
    value: '3',
  },
  {
    label: 'ABC',
    value: '4',
  },
]

describe('ScenarioLoaderList', () => {
  let onLoadButtonClick: (id: string) => void
  let onClose: () => void
  let wrapper: RenderResult

  let setSeverity: (severity: SeverityDistributionDatum[]) => void
  let scenarioDispatch: (action: AnyAction) => void

  beforeEach(() => {
    console.warn = noop
    console.error = noop

    onLoadButtonClick = jest.fn()
    onClose = jest.fn()
    setSeverity = jest.fn()
    scenarioDispatch = jest.fn()

    wrapper = render(
      <ScenarioLoader
        scenarioOptions={DATA}
        onLoad={onLoadButtonClick}
        onClose={onClose}
        visible
        setSeverity={setSeverity}
        scenarioDispatch={scenarioDispatch}
      />,
    )
  })

  afterEach(cleanup)

  it('matches snapshot', () => {
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders title', () => {
    expect(wrapper.getByText('Change scenario')).not.toBeNull()
  })

  it('displays scenarios', () => {
    const loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')

    expect(wrapper.getByText('Showing 4 of 4 entries')).not.toBeNull()
    expect(loadButtons).toHaveLength(4)
  })

  it('filters scenarios', async () => {
    const input = wrapper.getByTestId('PresetLoaderDialogInput')
    let loadButtons

    fireEvent.change(input, { target: { value: 'ABC' } })
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })

    await wait(() => wrapper.getByText('Showing 1 of 4 entries'))

    loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')
    expect(loadButtons).toHaveLength(1)

    const clearButton = wrapper.getByTestId('PresetLoaderDialogClearButton')
    fireEvent.click(clearButton)

    await wait(() => wrapper.getByText('Showing 4 of 4 entries'))

    loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')

    expect(loadButtons).toHaveLength(4)
  })

  describe('row button click triggers onLoadButtonClick handler', () => {
    it('when results not filtered', () => {
      const loadButton = wrapper.getAllByTestId('PresetLoaderDialogRowButton')

      fireEvent.click(loadButton[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('1')

      fireEvent.click(loadButton[1])

      expect(onLoadButtonClick).toHaveBeenCalledWith('2')
    })

    it('when results filtered', async () => {
      const input = wrapper.getByTestId('PresetLoaderDialogInput')

      fireEvent.change(input, { target: { value: 'ABC' } })
      fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })

      await wait(() => wrapper.getByText('Showing 1 of 4 entries'))

      const loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowButton')
      fireEvent.click(loadButtons[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('4')
    })
  })

  describe('row link click triggers onLoadButtonClick handler', () => {
    it('when results not filtered', () => {
      const loadButton = wrapper.getAllByTestId('PresetLoaderDialogRowLink')

      fireEvent.click(loadButton[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('1')

      fireEvent.click(loadButton[1])

      expect(onLoadButtonClick).toHaveBeenCalledWith('2')
    })

    it('when results filtered', async () => {
      const input = wrapper.getByTestId('PresetLoaderDialogInput')

      fireEvent.change(input, { target: { value: 'ABC' } })
      fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })

      await wait(() => wrapper.getByText('Showing 1 of 4 entries'))

      const loadButtons = wrapper.getAllByTestId('PresetLoaderDialogRowLink')
      fireEvent.click(loadButtons[0])

      expect(onLoadButtonClick).toHaveBeenCalledWith('4')
    })
  })

  it('close button triggers the onClose handler', () => {
    const closeButton = wrapper.getByTestId('PresetLoaderDialogCloseButton')
    fireEvent.click(closeButton)

    expect(onClose).toHaveBeenCalled()
  })
})
