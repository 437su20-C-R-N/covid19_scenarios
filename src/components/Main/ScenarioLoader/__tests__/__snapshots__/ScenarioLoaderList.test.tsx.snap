// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScenarioLoaderList matches snapshot 1`] = `<DocumentFragment />`;
